src/librustc_mir/transform/simplify_try.rs

```diff
@@ -577,3 +577,131 @@ impl<'tcx> MirPass<'tcx> for SimplifyBranchSame {
+                // we are looking for a basic block of only a single statement; discriminant(place) = VAR_INDEX
+                // this can be replaced with `place = discr_place` if the value of the switch statement matches the VAR_INDEX
+                if bbd.statements.len() != 1 {
+                    trace!("NO: Basic block is not just a single statement; discriminant(place) = VAR_INDEX");
+                    continue;
+                }
+                
+                let stmt = &bbd.statements[0];
+                match &stmt.kind {
+                    StatementKind::SetDiscriminant { place, variant_index } => {
+                        let place_type = place.ty(body, tcx).ty;
+                        let adt = match place_type.kind {
+                            ty::Adt(adt, _) if adt.is_enum() => adt,
+                            _ => {
+                                bug!("set descriminant on adt that is not enum")
+                            }
+                        };
+                        let variant_is_fieldless = adt.variants[*variant_index].fields.is_empty();
+                        
+                        if !variant_is_fieldless{
+                            trace!("NO: variant is not fieldless");
+                            continue; 
+                        } 
+                        let local = place.as_local();
+                        if local.is_none(){
+                            trace!("NO: place for enum is not a local variable");
+                            continue;
+                        } 
+                        let local = local.unwrap();
+                        let discr = adt.discriminant_for_variant(tcx, *variant_index);
+
+                        if body.local_decls()[discr_place.local].ty != body.local_decls()[local].ty {
+                            trace!("NO: types do not match");
+                        }
+
+                        // check if the discriminant value in the target bb matches with the discr being switched on
+                        let target_discr_value = values[target_idx];
+                        if discr.val == target_discr_value {
+                            // the opmization applies
+                            to_opt.push((bb_idx, Into::<_>::into(local.clone()), discr_place));
+                        } else {
+                            trace!("NO: discriminant on switch does not match target branch discriminant");
+                        }
+                    },
+                    _ => {}
+                }
+            }
+
+            
+        }
+        for (bb_with_first_stmt_to_overwrite, place, discr_place) in to_opt{
+            trace!("success: replacing set of discriminant with move");
+            body.basic_blocks_mut()[bb_with_first_stmt_to_overwrite].statements[0].kind = StatementKind::Assign(box (
+                place,
+                Rvalue::Use(Operand::Move(discr_place)),),
+            );
+        }
+    }
+}
```

Comment:
nit: missing trailing newline

Reply:
I don't think I have run rustfmt on the code yet